lib/cartopy/img_transform.py

```diff
@@ -302,38 +302,18 @@ def regrid(array, source_x_coords, source_y_coords, source_cs, target_proj,
 
     distances, indices = kdtree.query(target_xyz, k=1)
     mask = np.isinf(distances)
+    indices[mask] = 0
 
     desired_ny, desired_nx = target_x_points.shape
-    if array.ndim == 1:
-        if np.any(mask):
-            array_1d = np.ma.array(array[indices], mask=mask)
-        else:
-            array_1d = array[indices]
-        new_array = array_1d.reshape(desired_ny, desired_nx)
-    elif array.ndim == 2:
-        # Handle missing neighbours using a masked array
-        if np.any(mask):
-            indices = np.where(np.logical_not(mask), indices, 0)
-            array_1d = np.ma.array(array.reshape(-1)[indices], mask=mask)
-        else:
-            array_1d = array.reshape(-1)[indices]
-
-        new_array = array_1d.reshape(desired_ny, desired_nx)
-    elif array.ndim == 3:
-        # Handle missing neighbours using a masked array
-        if np.any(mask):
-            indices = np.where(np.logical_not(mask), indices, 0)
-            array_2d = array.reshape(-1, array.shape[-1])[indices]
-            mask, array_2d = np.broadcast_arrays(
-                mask.reshape(-1, 1), array_2d)
-            array_2d = np.ma.array(array_2d, mask=mask)
-        else:
-            array_2d = array.reshape(-1, array.shape[-1])[indices]
 
-        new_array = array_2d.reshape(desired_ny, desired_nx, array.shape[-1])
+    # Squash the first two dims of the source array into one
+    flat_array = array.reshape((-1,) + array.shape[2:])
+    if np.any(mask):
+        new_array = np.ma.array(flat_array[indices])
+        new_array[mask] = np.ma.masked
```

Comment:
Does passing `mask=mask` to the constructor not work here?

Reply:
No; `mask` will always be 1D here, but `flat_array` will still be multi-dimensional if `array` is at least 3D. I should probably change the name of `flat_array` since it's not necessarily 'flat'.